<!DOCTYPE html>
<html lang="ja-JP">
 <head>
  <meta charset="utf-8"/>
  <meta content="IE=edge" http-equiv="X-UA-Compatible"/>
  <meta content="width=device-width, initial-scale=1" name="viewport"/>
  <meta content="Sample Description" name="description"/>
  <meta content="Yuichi Ito" name="author"/>
  <title>
   Python入門
  </title>
  <link href="/img/favicon.ico" rel="shortcut icon"/>
  <!-- Bootstrap -->
  <link href="./assets/css/bootstrap.css" rel="stylesheet">
   <link href="./assets/css/bootstrap-theme.css" rel="stylesheet"/>
   <link href="./assets/css/bootstrap_u1.css" rel="stylesheet"/>
   <!-- Font -->
   <link href="https://fonts.googleapis.com/css?family=Roboto" rel="stylesheet"/>
   <link href="https://fonts.googleapis.com/earlyaccess/mplus1p.css" rel="stylesheet">
    <!-- Code Highlight -->
    <link href="./assets/css/codehilite_colorful.css" rel="stylesheet"/>
    <!-- Custom -->
    <link href="./assets/css/font-awesome.min.css" rel="stylesheet"/>
    <link href="./assets/icomoon/style.css" rel="stylesheet"/>
    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
   </link>
  </link>
 </head>
 <body>
  <div class="backgroundLayer">
   <!-- Fixed navbar -->
   <nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container">
     <div class="navbar-header">
      <button aria-controls="navbar" aria-expanded="false" class="navbar-toggle collapsed" data-target="#navbar" data-toggle="collapse" type="button">
       <span class="sr-only">
        Toggle navigation
       </span>
       <span class="icon-bar">
       </span>
       <span class="icon-bar">
       </span>
       <span class="icon-bar">
       </span>
      </button>
      <a class="navbar-brand " href="./">
       Python
      </a>
     </div>
     <div class="navbar-collapse collapse" id="navbar">
      <ul class="nav navbar-nav">
       <li class="">
        <a href="./navi.html">
         Navi
        </a>
       </li>
       <li class="">
        <a href="./qa.html">
         QA
        </a>
       </li>
       <li class="">
        <a href="./books.html">
         Books
        </a>
       </li>
       <li style="font-family: avenger">
        <a href="/">
         GC
        </a>
       </li>
      </ul>
     </div>
    </div>
   </nav>
   <div class="container">
    <div class="u1-header">
     <a href="./" style=" text-decoration: none;">
      <div class="u1-title">
       Python Boot Camp
      </div>
      <p class="lead u1-description">
       プログラミング言語Python訓練プログラム
      </p>
     </a>
    </div>
    <div class="row">
     <div class="col-sm-8 u1-main">
      <!-- GENERATED HTML START -->
      <!-- GENERATED HTML START -->
      <h1>
       プログラムのエラー
      </h1>
      <div class="no_bullets" id="toc_container">
       <p class="toc_title">
        本記事の内容
       </p>
       <ul class="toc_list">
        <li>
         <a href="#header-1">
          概要
         </a>
        </li>
        <li>
         <a href="header-2">
          プログラムの間違い
         </a>
        </li>
        <li>
         <a href="header-3">
          エラー
         </a>
         <ul>
          <li>
           <a href="#header-4">
            構文解析エラー
           </a>
          </li>
          <li>
           <a href="#header-5">
            実行時エラー
           </a>
          </li>
          <li>
           <a href="#header-6">
            検知されないエラー
           </a>
          </li>
         </ul>
        </li>
        <li>
         <a href="#header-7">
          スタックトレース
         </a>
        </li>
        <li>
         <a href="header-8">
          printデバッグ
         </a>
        </li>
        <li>
         <a href="header-9">
          代表的なエラー
         </a>
         <ul>
          <li>
           <a href="#header-10">
            宣言されていない変数や関数の利用
           </a>
          </li>
          <li>
           <a href="#header-11">
            リストの範囲外のアクセス
           </a>
          </li>
          <li>
           <a href="#header-12">
            関数の引数の数の間違い
           </a>
          </li>
          <li>
           <a href="#header-13">
            外部環境に依存するエラー
           </a>
          </li>
         </ul>
        </li>
       </ul>
      </div>
      <h2 id="header-1">
       概要
      </h2>
      <p>
       プログラムに誤りがあると、そこでエラーが発生して停止することがあります。
たとえばif文のコロン(:)を忘れたり、インデントの深さを間違うといった文法レベルの問題から、
0を持つ変数で割り算してしまうといった間違いです。
      </p>
      <p>
       間違いを直して正しいプログラムを作るには、
停止時のエラー内容である「スタックトレース」を呼んで問題箇所を把握し修正する必要があります。
      </p>
      <p>
       また、プログラムの文法や使い方自体は正しいものの、
自分が意図した通りにプログラムが動いていない場合もあります。
そのような場合はプログラムの動きを追って問題箇所を探すことが必要になります。
      </p>
      <h2 id="header-2">
       プログラムの間違い
      </h2>
      <p>
       数行、数十行のプログラムならともかく、ある程度の大きさのプログラムを書くと必ず間違いがあります。
間違いにもいろいろあり、Pythonが「
       <strong>
        エラー(例外)
       </strong>
       」として実行を停止してしまうものから、
Pythonのプログラムとしては間違っていないためエラーにはならいものの、
プログラマの意図通りに動いていないものなどがあります。
      </p>
      <p>
       本ページではそれらの間違いをどのように正すかについて扱います。
      </p>
      <p>
       なお、「発生することが期待されるエラー」というものがプログラミングにはあります。
たとえば外部サーバーから情報を取ってくるプログラムで、その相手のサーバーがダウンしている場合のエラーなどです。
そういったトラブルは「
       <strong>
        例外処理
       </strong>
       」と呼ばれる手法で対応しますが、それは例外処理のページにて解説します。
      </p>
      <h2 id="header-3">
       エラー
      </h2>
      <p>
       プログラムに文法的な間違いがあった場合に、
Pythonは「どこが、どういう理由で間違っている」ということを、エラーとして出力します。
      </p>
      <p>
       エラーは大まかに2つあり、「
       <strong>
        構文解析エラー(文法エラー)
       </strong>
       」と「
       <strong>
        実行時エラー
       </strong>
       」と呼ばれるものです。
      </p>
      <p>
       Pythonはモジュールを読み込んだ段階で、そのモジュールを上から下まで読み込み処理をします。
その際に「
       <strong>
        構文解析
       </strong>
       」という「どのようなプログラム構造になっているかの把握」をしています。
その構文解析中に問題が発生すると構文解析エラーが発生します。
Pythonの文法のトラブルですので、文法エラーともいえます。
      </p>
      <p>
       構文解析時に問題がなかったプログラムであっても、その実行時にエラーが発生することがあります。
実行時エラーと呼ばれるもので、例えば宣言されていない変数を参照しようとしたり、
0で割り算をした場合に発生します。
      </p>
      <h3 id="header-4">
       構文解析エラー
      </h3>
      <p>
       例えば以下のようなプログラムがあるとしましょう。
プログラムは「sample.py」というプログラムファイルに書かれています。
      </p>
      <h4>
       プログラム: 2行目に間違いがある
      </h4>
      <div class="codehilite">
       <pre><span></span><span class="k">print</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="k">if</span> <span class="bp">True</span>
  <span class="k">print</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre>
      </div>
      <p>
       このプログラムは2行目のif文に間違いがあります。
正しくは「if True:」となるべきところを、「:」を忘れています。
これはPythonの文法として誤りがあります。
      </p>
      <p>
       このプログラムを実行すると、以下のようなエラー出力が得られます。
      </p>
      <h4>
       コンソール: 実行するとエラーが発生
      </h4>
      <div class="codehilite">
       <pre><span></span>$ python3 sample.py
  File "sample.py", line 2
    if True
          ^
SyntaxError: invalid syntax
</pre>
      </div>
      <p>
       エラーの出力を上から順に確認すると、以下のことがわかります。
      </p>
      <ul>
       <li>
        sample.pyというファイルの2行目
       </li>
       <li>
        「if True」の後ろ側に問題がある
       </li>
       <li>
        「SyntaxError: invalid syntax」というエラーが発生
       </li>
      </ul>
      <p>
       どこでどう間違えているか出力しているため、修正箇所はひと目でわかります。
      </p>
      <p>
       エラー出力以外に着目してほしいのは「print(1)」が実行されていないことです。
エラー箇所は2行目にあるものの、1行目に間違いはありません。
上から順に実行されるのであれば、1行目は実行されて、実行結果に「1」と表示されそうな気がします。
      </p>
      <p>
       これには理由があり、Pythonは「構文解析を実行前に行う」という動きをしているためです。
モジュールが実行されるのは構文解析が終わった後なので、文法に誤りがあると実行する段階まで進めません。
      </p>
      <h3 id="header-5">
       実行時エラー
      </h3>
      <p>
       構文解析にパスをしたら、Pythonのプログラムが実行されます。
      </p>
      <p>
       実行される段階で発生したエラーは「実行時エラー」と呼ばれており、
そのプログラムファイルのなかのその行が実行されたタイミングでエラーが発生します。
      </p>
      <p>
       たとえば、以下の文法的には間違いがないプログラムがあるとします。
      </p>
      <div class="codehilite">
       <pre><span></span><span class="k">print</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="k">if</span> <span class="bp">True</span><span class="p">:</span>
  <span class="n">a</span> <span class="o">=</span> <span class="mi">5</span> <span class="o">/</span> <span class="mi">0</span>
  <span class="k">print</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre>
      </div>
      <p>
       このプログラムは文法的にはあっていますが、
3行目の「a = 5 / 0」が「0による割り算」なので間違っています。
      </p>
      <p>
       これを実行すると、以下のようなエラーが発生します。
      </p>
      <div class="codehilite">
       <pre><span></span>$ python3 sample.py
1
Traceback (most recent call last):
  File "sample.py", line 3, in &lt;module&gt;
    a = 5 / 0
ZeroDivisionError: division by zero
</pre>
      </div>
      <p>
       エラーの内容を読むと以下のことが分かります。
      </p>
      <ul>
       <li>
        sample.pyというファイルの3行目
       </li>
       <li>
        「a = 5 / 0」に問題がある
       </li>
       <li>
        「ZeroDivisionError: division by zero」というエラーが発生
       </li>
      </ul>
      <p>
       「Division」は割り算という意味なので、0による割り算エラーが発生しているということです。
      </p>
      <p>
       先ほどの文法エラーと頃なるのは出力の1行目に「1」という出力があることです。
これはプログラムの1行目にある「print(1)」が実行されているためです。
      </p>
      <p>
       実行時エラーはその行が実行されて始めてエラーとなるため、
その行にいたるまでの他の行は実行されています。
ただ、エラーが発生した以降の処理は出力に「print(2)」に相当する「2」がないことからも、
実行されていないことが分かります。
      </p>
      <h3 id="header-6">
       検知されないエラー
      </h3>
      <p>
       実行時エラーは実行されることで始めてエラーであると分かります。
逆にいえば「エラーがあっても、それが実行されないとエラーにならない」ということです。
      </p>
      <p>
       たとえば先ほどのプログラムを少し書き換えて、以下のようなコードとします。
      </p>
      <div class="codehilite">
       <pre><span></span><span class="k">print</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="k">if</span> <span class="bp">False</span><span class="p">:</span>
  <span class="n">a</span> <span class="o">=</span> <span class="mi">5</span> <span class="o">/</span> <span class="mi">0</span>
  <span class="k">print</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre>
      </div>
      <p>
       if文の条件式がFalseとなっているため、そのコードブロックは実行されません。
ただ、そのコードブロック中の処理にはエラーとなる「a = 5 / 0」があります。
      </p>
      <p>
       このプログラムを実行すると、以下のような出力となりエラーは発生しません。
      </p>
      <div class="codehilite">
       <pre><span></span>$ python3 sample3.py
1
3
</pre>
      </div>
      <p>
       実行時エラーは実行時にはじめて発生します。
そのため、条件分岐などの関係で「実行したり実行されなかったりする行の問題」は検知しづらい場合があります。
      </p>
      <p>
       実行時にエラーとなることを防ぐためには、きちんとした「
       <strong>
        テスト
       </strong>
       」が必要です。
ただ、初心者は他に優先して学ぶべきことが多いので、
ある程度のレベルに達するまではエラーが発生した際に直す姿勢で問題ないと思います。
一人前のコードが書けるようになったらテストなども気にかけて下さい。
      </p>
      <h2 id="header-7">
       スタックトレース
      </h2>
      <p>
       グローバルレベルで書かれている処理で実行時エラーが発生した場合は「何行目でエラー」と表示されるだけです。
ただ、関数や後述するクラスのメソッドでエラーが発生した場合は「どのように呼び出されたか」という
「
       <strong>
        スタックトレース
       </strong>
       」とともにエラーが表示されます。
      </p>
      <p>
       例えば以下のプログラムでは「function2」にバグがあり、
グローバルレベルからfunction1を経由してそれが呼び出されています。
      </p>
      <div class="codehilite">
       <pre><span></span><span class="k">def</span> <span class="nf">function1</span><span class="p">():</span>
  <span class="k">return</span> <span class="n">function2</span><span class="p">()</span>   <span class="c1"># line 2</span>

<span class="k">def</span> <span class="nf">function2</span><span class="p">():</span>
  <span class="k">return</span> <span class="mi">5</span><span class="o">/</span><span class="mi">0</span>           <span class="c1"># line 5</span>

<span class="n">function1</span><span class="p">()</span>            <span class="c1"># line 7</span>
</pre>
      </div>
      <p>
       これを実行すると、以下のようなスタックトレースが得られます。
      </p>
      <div class="codehilite">
       <pre><span></span>python3 sample.py
Traceback (most recent call last):
  File "sample.py", line 7, in &lt;module&gt;
    function1()
  File "sample.py", line 2, in function1
    return function2()
  File "sample.py", line 5, in function2
    return 5/0
ZeroDivisionError: division by zero
</pre>
      </div>
      <p>
       最終的に「0による割り算エラー」が発生してはいるものの、
「Traceback」に続いてエラーに至るまでの呼び出しの流れが上から下に順に記載されています。
      </p>
      <p>
       自分で作成した関数も様々な箇所から呼び出されます。
毎回エラーがでるのであれば対処も簡単ですが「X行目からの呼び出しは大丈夫」
「Y行目からの呼び出しはエラーになる」といった状況だと、
このスタックトレースを見ながら「どのような流れでエラーになったのか読み解く」という作業が必要になります。
      </p>
      <h2 id="header-8">
       printデバッグ
      </h2>
      <p>
       エラーの発生原因が複雑だと、問題箇所がすぐに特定できない場合があります。
そのような場合は「printデバッグ」と呼ばれる手法を使って調査ができます。
      </p>
      <p>
       たとえば「剰余」という「割り算の余り」の計算を使う「FizzBuzz」と呼ばれる有名なプログラムを組むとします。
ある範囲の数字(ここでは1から10とする)を2で割り切れるときはFizzと出力し、3で割り切れる時はBuzzと出力します。
そして2でも3でも割り切れる場合はFizzBuzzと出力します。
      </p>
      <p>
       剰余の計算は「%」記号を使います。
2で割り切れるということは、「整数 % 2」が0ということで、
3で割り切れるということは「整数 % 3」が0ということです。
      </p>
      <div class="codehilite">
       <pre><span></span>&gt;&gt;&gt; 5 % 2
1
&gt;&gt;&gt; 4 % 2
0
&gt;&gt;&gt; 5 % 3
2
&gt;&gt;&gt; 3 % 3
0
</pre>
      </div>
      <p>
       このfizzbuzzの間違いがあるコードを以下に示します。
      </p>
      <div class="codehilite">
       <pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">a</span><span class="p">:</span>
  <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'fizz'</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'buzz'</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">6</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'fizzbuzz'</span><span class="p">)</span>
</pre>
      </div>
      <p>
       2でも3でも割り切れるというのは6で割り切れるということなので、
「i%6 == 0」としています。
      </p>
      <p>
       このプログラムを実行すると、以下のようにfizzbuzzがないことが分かります。
      </p>
      <div class="codehilite">
       <pre><span></span>fizz
buzz
fizz
fizz
fizz
buzz
fizz
</pre>
      </div>
      <p>
       この原因を調べる時にprintデバッグを使うと便利です。
たとえば以下のようにコードを書き換えます。
      </p>
      <div class="codehilite">
       <pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">a</span><span class="p">:</span>
  <span class="k">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'fizz'</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'buzz'</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">6</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'fizzbuzz'</span><span class="p">)</span>
  <span class="k">print</span><span class="p">(</span><span class="s1">'----'</span><span class="p">)</span>
</pre>
      </div>
      <p>
       ループの中に何の値を調べているかを「print(i)」、
ループごとの区切りのを「print('fizzbuzz')」で表示するコードを追加いたしました。
      </p>
      <p>
       これを実行すると、以下のような結果が得られます。
      </p>
      <div class="codehilite">
       <pre><span></span>1
----
2
fizz
----
3
buzz
----
4
fizz
----
5
----
6
fizz
----
7
----
8
fizz
----
9
buzz
----
10
fizz
----
</pre>
      </div>
      <p>
       値6の箇所で「fizzbuzz」ではなく「fizz」と出力されていることがわかります。
if文の説明にて「elifはそれより前の条件が合致していた場合はチェックされない」としましたが、
今回は「elif i%6 == 0:」のチェックより前に「if i%2 == 0:」が合致してしまっているため、
fizzbuzzのチェックが実施されていません。
      </p>
      <p>
       問題箇所がわかったので、プログラムを修正できます。
      </p>
      <div class="codehilite">
       <pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">a</span><span class="p">:</span>
  <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">6</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'fizzbuzz'</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'fizz'</span><span class="p">)</span>
  <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">'buzz'</span><span class="p">)</span>
</pre>
      </div>
      <p>
       正しくfizzbuzzが表示されるようになりました。
      </p>
      <div class="codehilite">
       <pre><span></span>fizz
buzz
fizz
fizzbuzz
fizz
buzz
fizz
</pre>
      </div>
      <h2 id="header-9">
       代表的なエラー
      </h2>
      <p>
       Pythonのエラーの種類は豊富であり、自分で作ることさえできます。
それら全てを覚える必要はありませんが、よく見かけるエラーを紹介します。
      </p>
      <h3 id="header-10">
       宣言されていない変数や関数の利用
      </h3>
      <p>
       「
       <strong>
        名前エラー
       </strong>
       」は宣言されていない変数や関数を参照したときに発生するエラーです。
      </p>
      <div class="codehilite">
       <pre><span></span>&gt;&gt;&gt; a = 123
&gt;&gt;&gt; print(a)
123
&gt;&gt;&gt; print(b)
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
NameError: name 'b' is not defined
</pre>
      </div>
      <p>
       変数aは初期化されており、その参照は問題なくできています。
一方、bは特に宣言無く参照されているため、「bは定義されていない」と名前エラーが発生しています。
      </p>
      <p>
       よくあるのは変数名や関数名をタイプミスしている場合です。
たとえば「print('hello')」と書くつもりが「prinr('hello')」となっていれば、
prinrは存在しないので名前エラーが発生します。
      </p>
      <p>
       あまり大きなモジュールを作らずに適切に内部を構造化していれば、
問題がおきてもすぐに何を修正すればよいかが分かります。
      </p>
      <h3 id="header-11">
       リストの範囲外のアクセス
      </h3>
      <p>
       リストや文字列などはインデックス番号を指定して、その要素にアクセスできます。
ただ、その指定するインデックス番号がリストの範囲を越える場合は「
       <strong>
        インデックスエラー
       </strong>
       」が発生します。
      </p>
      <div class="codehilite">
       <pre><span></span>&gt;&gt;&gt; a = [1,2,3]
&gt;&gt;&gt; print(a[1])
2
&gt;&gt;&gt; print(a[100])
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
IndexError: list index out of range
</pre>
      </div>
      <p>
       変数aには長さが3のリストが格納されています。
指定できるのはインデックスの0から2までなので、100を指定すれば範囲外のアクセスとなります。
      </p>
      <p>
       Pythonのリストでループ処理をする場合は「for 変数 in リスト」を使うため、
インデックス番号の心配はいりません。
ただ、たまにインデックス番号を使う必要がある場面が発生するので、
そのような時は範囲外にアクセスしないように注意をしてください。
特にインデックス番号と長さの比較をする場合に「index &lt; X」とするか「index &lt;= X」とするかなどは、
よくミスをする箇所です。
      </p>
      <h3 id="header-12">
       関数の引数の数の間違い
      </h3>
      <p>
       関数を呼び出す際は引数を渡します。
関数が求める引数の数と、渡す引数の数が異なると「
       <strong>
        タイプエラー
       </strong>
       」が発生します。
      </p>
      <div class="codehilite">
       <pre><span></span>&gt;&gt;&gt; len([1,2,3])
3
&gt;&gt;&gt; len(1,2,3)
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
TypeError: len() takes exactly one argument (3 given)
</pre>
      </div>
      <p>
       len関数はリストや文字列などの長さを求める関数です。
この関数は引数を1つ取りますが、誤って3つを渡すとエラーになります。
      </p>
      <p>
       エラーにも「len関数は引数を1つとるが、3つ渡された」とあります。
      </p>
      <h3 id="header-13">
       外部環境に依存するエラー
      </h3>
      <p>
       Pythonのプログラムは外部環境にアクセスすることがあります。
たとえばPCにあるファイルを読み込んだり、ネットワークの機能を使ったりといった具合です。
      </p>
      <p>
       Pythonのプログラムとしては間違いがなくても、「読み込むファイルがなくなった」
「接続するサーバーがダウンしている」といった外部環境に依存して発生するトラブルは存在します。
これらは自分ではどうしようもない問題ですので、
「例外処理」というテクニックを使って「もしファイルがなかったらこうする」
「もしサーバーに繋がらなかったらこうする」といった具合でエラーが発生してもプログラムを継続できるようにします。
      </p>
      <div class="codehilite">
       <pre><span></span>&gt;&gt;&gt; import requests
&gt;&gt;&gt; requests.get('http://0.0.0.0/')
Traceback (most recent call last):
...省略...
ConnectionRefusedError: [Errno 61] Connection refused
</pre>
      </div>
      <!-- GENERATED HTML END -->
      <!-- GENERATED HTML END -->
      <nav>
       <ul class="pager">
        <li id="previous-li">
         <a href="./018.html" id="previous-a">
          ← Previous
         </a>
        </li>
        <li>
         <a href="#">
          Page Top
         </a>
        </li>
        <li class="pager-next" id="next-li">
         <a href="./020.html" id="next-a">
          Next →
         </a>
        </li>
       </ul>
      </nav>
     </div>
     <!-- /.blog-main -->
     <div class="col-sm-3 col-sm-offset-1 u1-sidebar">
      <!-- SNS start -->
      <div class="sidebar-module">
       <ul class="social-buttons">
        <li>
         <!-- hatena -->
         <a class="hatena-bookmark-button" data-hatena-bookmark-lang="ja" data-hatena-bookmark-layout="basic-label-counter" href="http://b.hatena.ne.jp/entry/" title="このエントリーをはてなブックマークに追加">
          <img alt="このエントリーをはてなブックマークに追加" height="20" src="https://b.st-hatena.com/images/entry-button/button-only@2x.png" style="border: none;" width="20"/>
         </a>
         <script async="async" charset="utf-8" src="https://b.st-hatena.com/js/bookmark_button.js" type="text/javascript">
         </script>
        </li>
        <li>
         <!-- FB -->
         <div class="fb-like" data-action="like" data-layout="button_count" data-share="true" data-show-faces="false" data-size="small">
         </div>
         <div id="fb-root">
         </div>
         <script>
          (function(d, s, id) {
                    var js, fjs = d.getElementsByTagName(s)[0];
                    if (d.getElementById(id)) return;
                    js = d.createElement(s);
                    js.id = id;
                    js.src = "//connect.facebook.net/ja_JP/sdk.js#xfbml=1&version=v2.9";
                    fjs.parentNode.insertBefore(js, fjs);
                  }(document, 'script', 'facebook-jssdk'));
         </script>
        </li>
        <li>
         <!-- twitter -->
         <a class="twitter-share-button" href="https://twitter.com/share">
          Tweet
         </a>
         <script>
          ! function(d, s, id) {
                    var js, fjs = d.getElementsByTagName(s)[0],
                      p = /^http:/.test(d.location) ? 'http' : 'https';
                    if (!d.getElementById(id)) {
                      js = d.createElement(s);
                      js.id = id;
                      js.src = p + '://platform.twitter.com/widgets.js';
                      fjs.parentNode.insertBefore(js, fjs);
                    }
                  }(document, 'script', 'twitter-wjs');
         </script>
        </li>
        <li>
         <!-- G+ -->
         <!-- head 内か、body 終了タグの直前に次のタグを貼り付けてください。 -->
         <script async="" defer="" src="https://apis.google.com/js/platform.js">
          {
                    lang: 'ja'
                  }
         </script>
         <!-- 共有ボタン を表示したい位置に次のタグを貼り付けてください。 -->
         <div class="g-plus" data-action="share" data-annotation="none" data-height="24">
         </div>
        </li>
        <li>
         <!-- IN/Share -->
         <script src="//platform.linkedin.com/in.js" type="text/javascript">
          lang: ja_JP
         </script>
         <script data-counter="right" type="IN/Share">
         </script>
        </li>
       </ul>
      </div>
      <!-- SNS end -->
      <div class="sidebar-module">
       <p class="sidebar-heading">
        Python入門
       </p>
       <ol class="list-unstyled">
        <li>
         <a href="./">
          トップページ
         </a>
        </li>
        <li>
         <a href="./navi.html">
          目次
         </a>
        </li>
        <li>
         <a href="./qa.html">
          質問と回答
         </a>
        </li>
        <li>
         <a href="./books.html">
          おすすめPython書籍
         </a>
        </li>
       </ol>
      </div>
      <div class="sidebar-module">
       <p class="sidebar-heading">
        Pythonの基礎
       </p>
       <ol class="list-unstyled">
        <li>
         <a href="./001.html">
          プログラミングとは
         </a>
        </li>
        <li>
         <a href="./002.html">
          なぜPythonを学ぶのか
         </a>
        </li>
        <li>
         <a href="./003.html">
          Python3のインストール
         </a>
        </li>
        <li>
         <a href="./004.html">
          インタプリタの実行
         </a>
        </li>
        <li>
         <a href="./005.html">
          プログラムファイルの実行
         </a>
        </li>
        <li>
         <a href="./006.html">
          コメント
         </a>
        </li>
        <li>
         <a href="./007.html">
          データと型
         </a>
        </li>
        <li>
         <a href="./008.html">
          変数
         </a>
        </li>
        <li>
         <a href="./009.html">
          変数の名前付け
         </a>
        </li>
        <li>
         <a href="./010.html">
          定数とマジックナンバー
         </a>
        </li>
        <li>
         <a href="./011.html">
          主要な型と演算子
         </a>
        </li>
        <li>
         <a href="./012.html">
          条件分岐とコードブロック
         </a>
        </li>
        <li>
         <a href="./013.html">
          ループ
         </a>
        </li>
        <li>
         <a href="./014.html">
          関数
         </a>
        </li>
        <li>
         <a href="./015.html">
          名前空間
         </a>
        </li>
        <li>
         <a href="./016.html">
          オブジェクト指向を体験
         </a>
        </li>
        <li>
         <a href="./017.html">
          モジュールの利用
         </a>
        </li>
        <li>
         <a href="./018.html">
          モジュールの作成
         </a>
        </li>
        <li>
         <a href="./019.html">
          プログラムのエラー
         </a>
        </li>
       </ol>
      </div>
      <div class="sidebar-module">
       <p class="sidebar-heading">
        Pythonの基本的な型
       </p>
       <ol class="list-unstyled">
        <li>
         数値型
        </li>
        <li>
         文字列型
        </li>
        <li>
         ブール型
        </li>
        <li>
         None型
        </li>
        <li>
         <a href="./055.html">
          リスト型
         </a>
        </li>
        <li>
         <a href="./056.html">
          タプル型
         </a>
        </li>
        <li>
         レンジ型
        </li>
        <li>
         バイト型
        </li>
        <li>
         <a href="./059.html">
          セット型
         </a>
        </li>
        <li>
         <a href="./060.html">
          辞書型
         </a>
        </li>
       </ol>
      </div>
      <!-- amazon -->
      <div class="sidebar-module">
       <script type="text/javascript">
        amazon_ad_tag = "grit-22";
              amazon_ad_width = "300";
              amazon_ad_height = "250";
              amazon_ad_link_target = "new";
       </script>
       <script src="http://ir-jp.amazon-adsystem.com/s/ads.js">
       </script>
      </div>
      <!-- amazon end-->
      <div class="sidebar-module">
       <p class="sidebar-heading">
        Python入門
       </p>
       <ol class="list-unstyled">
        <li>
         <a href="https://github.com/yuichi110/www.yuichi.com-python" target="_blank">
          <span class="fa fa-github">
          </span>
          GitHub Repository
         </a>
        </li>
        <li>
         <a href="https://twitter.com/yuichi110" target="_blank">
          <i class="fa fa-twitter">
          </i>
          Twitter
         </a>
        </li>
        <li>
         <a href="https://www.linkedin.com/in/yuichiito/" target="_blank">
          <span class="fa fa-linkedin">
          </span>
          Linkedin (英語)
         </a>
        </li>
        <li class="li-none">
         <a href="#" id="rw_email_contact">
          <span class="fa fa-envelope">
           Mail
          </span>
         </a>
         <script type="text/javascript">
          var _rwObsfuscatedHref0 = "mai";
                  var _rwObsfuscatedHref1 = "lto";
                  var _rwObsfuscatedHref2 = ":yu";
                  var _rwObsfuscatedHref3 = "ich";
                  var _rwObsfuscatedHref4 = "i@y";
                  var _rwObsfuscatedHref5 = "uic";
                  var _rwObsfuscatedHref6 = "hi.";
                  var _rwObsfuscatedHref7 = "com";
                  var _rwObsfuscatedHref = _rwObsfuscatedHref0 + _rwObsfuscatedHref1 + _rwObsfuscatedHref2 + _rwObsfuscatedHref3 + _rwObsfuscatedHref4 + _rwObsfuscatedHref5 + _rwObsfuscatedHref6 + _rwObsfuscatedHref7;
                  document.getElementById("rw_email_contact").href = _rwObsfuscatedHref;
         </script>
        </li>
       </ol>
      </div>
      <div class="sidebar-module">
       <!-- padding for side bar -->
       <div class="hidden-xs">
        <p style="height:500px">
        </p>
       </div>
      </div>
     </div>
     <!-- /.blog-sidebar -->
    </div>
    <!-- /.row -->
   </div>
   <!-- /.container -->
   <footer>
    <div class="u1-footer-top">
    </div>
    <div class="u1-footer">
     <p>
      Copyright © 2017 Yuichi Ito All Rights Reserved.
     </p>
     <p align="center">
      修正依頼がある場合は右ページよりページURLと問題箇所を指摘願います。
      <a href="https://github.com/yuichi110/www.yuichi.com-python/issues">
       <span class="fa fa-github">
       </span>
       Git Hub Issues
      </a>
     </p>
    </div>
   </footer>
   <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js">
   </script>
   <script src="./assets/js/bootstrap.min.js">
   </script>
   <script src="./assets/js/delete_useless_dom.js">
   </script>
   <!-- GOOGLE ANALYTICS -->
   <script>
    (function(i, s, o, g, r, a, m) {
        i['GoogleAnalyticsObject'] = r;
        i[r] = i[r] || function() {
          (i[r].q = i[r].q || []).push(arguments)
        }, i[r].l = 1 * new Date();
        a = s.createElement(o),
          m = s.getElementsByTagName(o)[0];
        a.async = 1;
        a.src = g;
        m.parentNode.insertBefore(a, m)
      })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
      ga('create', 'UA-4830860-7', 'auto');
      ga('send', 'pageview');
   </script>
  </div>
  <!-- end of background -->
 </body>
</html>
